upf/transformers/grounder.py

```diff
@@ -70,29 +85,39 @@ def get_rewritten_problem(self) -> Problem:
+                if self._grounding_actions_map is None or \
+                    self._grounding_actions_map.get(old_action, None) is not None:
+                    new_action = old_action.clone()
+                    self._new_problem.add_action(new_action)
+                    self._new_to_old[new_action] = old_action
+                    self._map_parameters[new_action] = []
+                    self._old_to_new[old_action] = [new_action]
                 continue
-            # a list containing the list of object in the problem of the given type.
-            # So, if the problem has 2 Locations l1 and l2, and 2 Robots r1 and r2, and
-            # the action move_to takes as parameters a Robot and a Location,
-            # the variable state at this point will be the following:
-            # type_list = [Robot, Location]
-            # objects_list = [[r1, r2], [l1, l2]]
-            # the product of *objects_list will be:
-            # [(r1, l1), (r1, l2), (r2, l1), (r2,l2)]
-            ground_size = 1
-            domain_sizes = []
-            for t in type_list:
-                ds = domain_size(self._new_problem, t)
-                domain_sizes.append(ds)
-                ground_size *= ds
-            items_list: List[List[FNode]] = []
-            for size, type in zip(domain_sizes, type_list):
-                items_list.append([domain_item(self._new_problem, type, j) for j in range(size)])
-            for grounded_params in product(*items_list):
+            grounded_params_list: Optional[Iterator[Tuple[FNode, ...]]] = None
+            if self._grounding_actions_map is None:
+                # a list containing the list of object in the problem of the given type.
+                # So, if the problem has 2 Locations l1 and l2, and 2 Robots r1 and r2, and
+                # the action move_to takes as parameters a Robot and a Location,
+                # the variable state at this point will be the following:
+                # type_list = [Robot, Location]
+                # objects_list = [[r1, r2], [l1, l2]]
+                # the product of *objects_list will be:
+                # [(r1, l1), (r1, l2), (r2, l1), (r2,l2)]
+                ground_size = 1
+                domain_sizes = []
+                for t in type_list:
+                    ds = domain_size(self._new_problem, t)
+                    domain_sizes.append(ds)
+                    ground_size *= ds
+                items_list: List[List[FNode]] = []
+                for size, type in zip(domain_sizes, type_list):
+                    items_list.append([domain_item(self._new_problem, type, j) for j in range(size)])
+                grounded_params_list = product(*items_list)
+            else:
+                # The grounding_actions_map is not None, therefore it must be used to ground
+                grounded_params_list = _create_generator_from_list_of_tuple(self._grounding_actions_map[old_action])
```

Comment:
There is no need for this function:
```suggestion
                grounded_params_list = iter(self._grounding_actions_map[old_action])
```